data_tools/firebase_client_wrapper.py

```diff
@@ -132,11 +170,43 @@ def delete_dataset(dataset_id):
 def _delete_collection(coll_ref, batch_size):
     docs = coll_ref.limit(batch_size).get()
     deleted = 0
-
     for doc in docs:
         print(u'Deleting doc {} => {}'.format(doc.id, doc.to_dict()))
         doc.reference.delete()
         deleted = deleted + 1
 
     if deleted >= batch_size:
         return _delete_collection(coll_ref, batch_size)
+
+def get_shard_count(dataset_id):
+    shard_count_doc = client.document(f'shard_counts/{dataset_id}').get().to_dict()
+    if shard_count_doc is None:
+        return None
+    return shard_count_doc["shard_count"]
+
+def set_shard_count(dataset_id, shard_count):
+    client.document(f'shard_counts/{dataset_id}').set({"shard_count": shard_count})
+
+def create_next_dataset_shard(dataset_id):
+    shard_count = get_shard_count(dataset_id)
+
+    if shard_count is None:
+        current_shard_id = f"{dataset_id}"
+        next_shard_id = f"{dataset_id}_2"
+        next_shard_count = 2
+    else:
+        current_shard_id = f"{dataset_id}_{shard_count}"
+        next_shard_id = f"{dataset_id}_{shard_count + 1}"
+        next_shard_count = shard_count + 1
+
+    print(f"Creating next dataset shard with id {next_shard_id}")
+
+    code_schemes = get_all_code_schemes(current_shard_id)
+    set_all_code_schemes(next_shard_id, code_schemes)
+
+    users = get_user_ids(current_shard_id)
+    set_users(next_shard_id, users)
+
+    set_shard_count(dataset_id, next_shard_count)
+
```

Comment:
I think here it would be advisable to put in a loop that reads get_shard_count, and tests that the write has committed, and if not, waits. Repeat 10 times and then crash.

Reply:
added